pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,179 @@
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search)
+		if err != nil {
+			// no match on search returns an error. do not log
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &userAccounts); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, u := range userAccounts {
+			u.Type = UserType
+			accounts = append(accounts, u)
+		}
+	}
+	if principalType == "" || principalType == GroupType {
+		var groups []Group
+		searchURL := fmt.Sprintf("%s/%ss?search=%s", k.getSearchURL(config.Issuer), GroupType, searchTerm)
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search)
+		if err != nil {
+			// no match on search returns an error. do not log
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &groups); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, g := range groups {
+			accounts = append(accounts, Account{ID: g.ID, Name: g.Name, Type: GroupType})
+			subGroups := getSubGroups(g)
+			for _, sg := range subGroups {
+				accounts = append(accounts, Account{ID: sg.ID, Name: sg.Name, Type: GroupType})
+			}
+		}
+	}
+	return accounts, nil
+}
+
+func getSubGroups(group Group) []Group {
+	var groups []Group
+	if len(group.Subgroups) > 0 {
+		for _, sub := range group.Subgroups {
+			groups = append(groups, sub)
+			groups = append(groups, getSubGroups(sub)...)
+		}
+	}
+	return groups
+}
+
+func (k *KClient) keycloakGroupSearch(searchTerm, searchType string, accessToken string, config *v32.OIDCConfig) ([]Group, error) {
+	searchURL := fmt.Sprintf("%s/%s?search=%s", k.getSearchURL(config.Issuer), searchType, searchTerm)
+	search := URLEncoded(searchURL)
+
+	b, err := k.getFromKeyCloak(accessToken, search)
+	if err != nil {
+		// no match on search returns an error. do not log
+		return nil, nil
+	}
+	var searchResults []Group
+	if err := json.Unmarshal(b, &searchResults); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResults, err
+	}
+	return searchResults, nil
+}
+
+func (k *KClient) getFromKeyCloakByID(principalID, accessToken, searchType string, config *v32.OIDCConfig) (Account, error) {
+	searchURL := fmt.Sprintf("%s/%s/%s", k.getSearchURL(config.Issuer), searchType, principalID)
+	search := URLEncoded(searchURL)
+	b, err := k.getFromKeyCloak(accessToken, search)
+	if err != nil {
+		return Account{}, nil
+	}
+	var searchResult Account
+	if err := json.Unmarshal(b, &searchResult); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResult, err
+	}
+	return searchResult, nil
+
+}
+func (k *KClient) getSearchURL(issuer string) string {
+	splitIssuer := strings.SplitAfter(issuer, "/auth/")
```

Comment:
Do we have validation that issuer always has this format ?

Reply:
I added validation that the issuer was in expected format prior to splitting it to ensure the search url function works as expected. since it is keycloak specific search we can expect the issuer to always be in this specific format.

Reply:
This would fail before create happens, so we don't need to check here.